### ISSUE TITLE
What to do about conditionals

### ISSUE THREAD

**Issue by mrfigg** [Morgul/rpgdice#8]
Starting a separate issue to further talk about proposed change 1 in #7.

**mrfigg:**
>     1. Add logical support
>        
>        1. Add `if(conditionExpr, thenExpr[, elseExpr])` function to default scope, internally it would use `if(conditionExpr.value !== 0) { value = thenExpr.value } else { value = /* elseExpr.value or 0 */ }`
>        2. Add `equal`, `notEqual`, `greaterThan`, `lessThan`, `greaterThanOrEqualTo`, `lessThanOrEqualTo`, `or`, `and`, & `not` operators
>           This can be done in two ways;
>           
>           1. Via functions in default scope `if(greaterThanOrEqualTo(1d20+5, enemyArmor), 2d8+3)`
>              This is simpler to implement, but not by much, and it is uglier
>           2. Via operators in peg file `if(1d20+5 >= enemyArmor, 2d8+3)`
>              The `not` operator (`!expr`) probably needs its own expression, but the rest would fit nicely in the `Operation` expression
>           
>           I would pick option b, but both are reasonable.

**Morgul:**
>     1. My first thought was, "If I've implemented a Turing complete Domain Specific Language for my dice roller, I'm doing something wrong." Now, that's not _entirely_ true; but it does feel like you're trying to put game specific logic into a DSL, when more than likely it should be implemented in side a 'real' language, like Javascript. So as far as `if`, `else`, `not`, `and`, `or` go, I'm rather against it. _However_, all of that being said, I'm open to the idea of expanding the parser to support user extensions for these use cases. Here's a couple of ideas:
>        
>        1. **Compromise:** Support the `?` operator.
>           
>           1. Yes, I know what I said above, but I can definitely see some places where something like `attackBonus > 5 ? 2d8 + 1 : 2d6 + 1` could be really handy. However, it would have the following caveats: it would use the short form javascript ternary, to discourage chained usage, and only `<, >, =` (and combinations) would be implemented, probably as their own `BooleanOperation` class. There would be _no_ `and`, `or`, `is`, or `not` implemented as operators. _(More on that below.)_
>           2. We _might_ need to implement `!`, and combinations like `!=`, depending on how awkward it is to use `not` with one of the other proposals below. We would **not** implement `===` or `!==`, instead `==` would be equivalent to `===` and `!=` would be equivalent to `!==`. (I do not want javascript's weird implicit casting part of this.).
>        2. **Implicit / Operator functions:** We could make the following valid: `not 3d6` where `not` is a function on the scope that takes one argument, basically `not 3d6 === not(3d6)`. _I have not thought through the parser implications of this!_ It would get us much closer to being able to have user defined operator functions like: `1d10 > 5 and 3d6 < 5`. In order to do that, however, we have to make a few definitions:
>           
>           1. An "implicit" function (I'd rather call them 'operator functions') could only be of arity one, or arity two.
>           2. If an operator function is of arity one, its signature would be: `op(right)`.
>           3. If an operator function is of arity two, its signature would be: `op(right, left)`.
>           4. If `Variable` evaluates to a function, it will be called with `fun(right, left)`, where `left` could be undefined.
>           5. I'm honestly not sure that the complexity of this is worth it. Even as I was typing this out I became concerned.
>        3. **Pipelines:** We could implement the minimal pipeline operator instead of the implicit operators. It would mean the following would be valid: `attackBonus |> greaterThan(3) ? 3d8 : 2d6`, where `greaterThan` was defined as `greaterThan(n) { return (m) => n > m; }`. This would also likely mean we'd need to ensure we support currying: `add(2)(3) === add: n => m => n + m`. I would also be open to a `lodash.fp.flow` way of doing custom operations: `flow(double(3(5d6 + 1)), negate, add(30))`. That's probably overkill, but as long as we support currying, a user could implement that.
>        4. **RPGDice Extras:** If we have all the base required support in the main lib, I'd be open to adding some of the more controversial features into a secondary library. It would, basically, be an expanded default scope, with all the function versions of the features you were suggesting. That way, if someone really wants those, they can have a nice, supported version, but for people who just want to roll some dice (my main use case), we keep the core lib nice and small.

**Comment by Morgul** [comment]
I started to formalize the `case ... of` proposal, but I ended up realizing it doesn't work unless you can declare variables inside of the grammar, and I'd like to avoid that. Instead, I'm going to start with a more traditional proposal...

### If/Then/Else

```
Conditional: Expression BooleanOp Expression [...BooleanOp]

if Conditional then Expression else Expression
```

Example:
```elm
if 3d6 + 1 < 3 then 2d6 + 1 else 3d8 + 12
```

A `Conditional`, for this discussion, always evaluates to a `Boolean`, and is the result of a `BooleanOp` (ex: `and`, `or`, `<`, `>=`, `!=`, etc.) A `BooleanOp`, on the other hand, is an operator that always returns a `Boolean`. They are not allowed outside of the conditional part of an `if ... then ... else` statement.

Some considerations: without making the `else` required, we would have to decide how the entire `if ... then` evaluates in the case the condition isn't met. Additionally, where in the grammar would we allow conditionals? Only top level? Anywhere? Things could get _very_ complicated depending on what we decide.

Given these, I propose the following rules:

* The `else` block **must always** be present.
* The `Conditional` **must always** be a Boolean. _(Casts are handled by the `BooleanOp`)_
* The `Expression` portion of the `then` and `else` blocks **must always** evaluate to a `Number`.
* Any `BooleanOp`s **must always** have well defined casting:
    * `0 -> false` and `true = !false`. _(Any number other than `0` is true.)_
* `if ... then ... else` **must always** be allowed where ever an `Expression` is allowed by the Grammar.
* `if ... then ... else` **must not** be evaluated by `Variable`.

#### Examples

```elm
if attackBonus > 5 then
    2 * (3d8 + [Weapons.Longsword.Damage])
else
   3d8 + [Weapons.Longsword.Damage]
```

```elm
if if if if 1d6 > 1 then 2d6 else 3 then 1d6 +1 else -7 then 3(3(3d12 + 7) + 9) else 1d2 + 1 then 2 else 3
```

```elm
3d10 + [Attrs.StrMod] + if [Attrs.DexMod] > 2 then [Attrs.DexMod] else 0
```

#### Additional Thoughts

Seeing this on paper, I don't hate it nearly as much. I think, if we implement this, I would like to also support the ternary `Condition ? Expression : Expression` syntax, since it's more succinct, and should follow exactly the same rules as `if ... then ... else`.

**Comment by Morgul** [comment]
Here's another proposal that, I think we could support without changes to the grammar.

### Conditional Functions

```
Conditional: Expression BooleanOp Expression [...BooleanOp]
ThenExpr: Expression
ElseExpr: Expression

if(Condition, ThenExpr, ElseExpr)
```

Example:
```javascript
if(3d6 + 1 < 3, 2d6 + 1, 3d8 + 12)
```

The `Conditional` portion would be cast to a `Boolean`, no matter what is passed in, meanwhile `ThenExpr` and `ElseExpr` would throw a `TypeError` if it doesn't return a `Number`.

A basic implementation of this could be:

```javascript
function if(cond, thenExpr, elseExpr)
{
    cond = cond.eval();
    if(cond.value === 0)
    {
        return thenExpr.eval();
    } // end if

    return elseExpr.eval();
} // end if
```

I _think_ that would work the way we want, and should be pretty easy to test with. While I'm not sold on  how this feels as a part of the syntax, this could be tenable. I would still want to implement the different `BooleanOp`s from above.

#### Examples

```javascript
if(
    attackBonus > 5, 
    2 * (3d8 + [Weapons.Longsword.Damage]), 
    3d8 + [Weapons.Longsword.Damage]
)
```
```javascript
if( if( if( if( 1d6 > 1, 2d6,  3), 1d6 +1, -7) 3(3(3d12 + 7) + 9), 1d2 + 1), 2, 3)
```

```javascript
3d10 + [Attrs.StrMod] + if([Attrs.DexMod] > 2, [Attrs.DexMod], 0)
```

**Comment by Morgul** [comment]
I've almost convinced myself of my first proposal, to be honest. The function version is ugly.

**Comment by mrfigg** [comment]
I'm in favor of your **If/Then/Else** proposal and am willing to help implement it as is, I'm just going to have to put some thought into how to implement the peg parse tree. I think it's going to require two different major branches, one simpler branch for handling only math and one expanded branch for handling both `BooleanOp`s and math. Also, just to float the idea, should we only use `... ? ... : ...` and not `if ... then ... else` as they are functionally the same, and the readme seems to indicate this module is going for a javascript-like syntax rather than an, I don't know..., python-like one? And we wouldn't have to worry about any confusion between the conditional and variables/functions. Just a thought/personal preference, feel free to ignore.

### Alternatively... preemptive type coercion

There is also an alternative to implementing **If/Then/Else** as is that, while a bit unorthodox, would allow us to keep a unified parse tree. I'm not sure I am going to be able to explain this one well, so I apologize in advance. We could have every op that would return a boolean preemptively use type coercion to return a number instead, because it's the only value type we want to be dealing with anyway. Then every op that needs to test for a boolean would check `if(value !== 0)` for true or `if(value === 0)` for false. This means these ops won't break any math if they are used unexpectedly outside of a conditional.

#### Examples

Given that `>=` would be:
```js
{
  greaterThanOrEqual:
  {
    symbol: '>=',
    evalValue: (left, right) =>
    {
      return left.value >= right.value ? 1 : 0;
    }
  }
}
```

And that `... ? ... : ...` would be something along the lines of:
```js
eval(scope, depth = 1)
{
  scope = defaultScope.buildDefaultScope(scope);

  this.condition = this.condition.eval(scope, depth + 1);
  this.then = this.then.eval(scope, depth + 1);
  this.else = this.else.eval(scope, depth + 1);

  if(this.condition.value !== 0)
  {
    this.value = this.then.value;
  }
  else
  {
    this.value = this.else.value;
  }
}
```

We can handle the expected use:
`(1d20 + 5) >= enemyArmor ? 2d8 + 3 : 0`

As well as:
`max(1d100 - chancePenalty, 0) ? 2d8 + 3 : 0`

And then even be able to handle someone doing something unexpected:
`((1d20 + 5) >= enemyArmor) * (2d8 + 3)`

I admit that last one is not very intuitive, but I'd argue the input for that last one isn't intuitive either, so it's not our fault. At least we're still able to handle it. And this should still feel very similar to javascript as you can do some weird things like `true * 5 = 5`, `false * 5 = 0`, and `true + true + false = 2` already anyway.

I do feel this idea could still be pretty out there, so I understand if it's too far of a stretch.

**Comment by mrfigg** [comment]
Actually now that I've slept on it I realize it should be possible to leave the `BooleanOp`s in the normal peg tree and to just let javascript handle type coercion natively. This would mean that a user could get a root expression with a boolean value if they did `rpgdice.eval('1d20 + 5 >= enemyArmor', {...})`, but that's what they're asking for.

Rereading your proposal I guess I just don't see the advantage of these two points;
>* The `Conditional` **must always** be a Boolean. _(Casts are handled by the `BooleanOp`)_
>* The `Expression` portion of the `then` and `else` blocks **must always** evaluate to a `Number`.

Ultimately I'm arguing in favour of the flexibility javascript type coercion allows.

**Comment by Morgul** [comment]
Responding in order:

### Ternary vs IfThenElse

I kinda want both, `if ... then ... else` and `... ? ... : ...`, but for what might be a silly reason; I *hate* the look of multiline expressions with ternary:

```javascript
attackBonus > 5 
    ? 2 * (3d8 + [Weapons.Longsword.Damage])
    : 3d8 + [Weapons.Longsword.Damage]
```

But, this is both not javscript, and involves more typing:

```elm
if attackBonus > 5 then 2 * (3d8 + [Weapons.Longsword.Damage]) else 3d8 + [Weapons.Longsword.Damage]
```

(In fact, I've lifted `if ... then ... else ...` from Elm; it's syntax is basically exactly what we wanted, it felt like.)

At the end of the day, `?` might be more consistent; as long as you think it'll work for your use case, I do prefer limiting the nesting, or some of the crazier use cases like that.

#### Preemptive Type Coercion

This isn't as weird as you might have thought; one of the gotchas about everything here is that right now we only have one output type: Numbers. Booleans will need to have some sort of casting to them otherwise people can accidentally do illegal things that will (_should_) blow up with a TypeError. As an approach, this isn't terrible.

#### Javascript Coercion

Regarding my points:

> * The `Conditional` **must always** be a Boolean. _(Casts are handled by the `BooleanOp`)_
> * The `Expression` portion of the `then` and `else` blocks **must always** evaluate to a `Number`.

My reasoning goes like this: We only need `Booleans` to exist for conditionals. I suppose this should be relaxed to "The `Conditional` **may be** a `Boolean`." I'm totally find with javascript coercion here.

The part that I'm not cool with is `rpgdice.eval('1d20 + 5 >= enemyArmor', {...})`. That should, IMHO throw a `TypeError` with a helpful message, like, "Expression does not evaluate to a numeric." What I want is to prevent mixing `Boolean` and `Numeric` operations in awkward ways.

That being said, Javascript does seem to have some decent rules for this:
```
> true + 1
2
> false + 1
1
> 50 / true
50
> 50 / false
Infinity
> parseFloat(true)
NaN
> parseInt(true)
NaN
> true + 0
1
> 
```
I just don't want the end result of the evaluation to be `true` or `false`. It should be a number, always.

**Comment by mrfigg** [comment]
**Ternary vs IfThenElse**
It seems we both just have personal preferences on this one, both of which I think are valid, and neither of which should really get in the way of the other. So I say let's just implement both and let the user decide which they find most appealing.

Could you clarify what you mean by _limiting the nesting_? I believe they should also be right-to-left associative as per [Operator Precedence](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence).

**No top level `BooleanOp`s**
Thinking on it not allowing top level booleans makes complete sense, but in order to accomplish this and truly protect against odd behavior I think we might be forced into going with **preemptive type coercion**. Otherwise what's to stop someone from trying to do this;
```js
let expr = rpgdice.parse('1d20 + 5 >= enemyArmor ? 0 : 0');
expr = expr.condition;
expr.eval({...});
```
The `condition` has no way to know if it's been pulled out of its `conditional`. Or should we consider this to be beyond our control and let the user do what they want?

And with **preemptive type coercion** a user could still do `rpgdice.eval('1d20 + 5 >= enemyArmor', {...})` if they really want, and still have the output be usable for their (probable) intended purpose and be a valid number.

**Comment by Morgul** [comment]
#### Limited Nesting

All I meant was, it gets really ugly trying to do:

```javascript
5 >= armorBonus ? 3 < armorBonus ? 9 > dexBonus ? 3d6 : 2d6 : 9 : 0
```

Which means the user won't try to do it too much. No actual enforcement requires, and it should definitely follow Operator Precedence.

For now, let's just go with ternary. We can always add `if ... then ... else` support later, if people start complaining about formatting nested ternaries.

#### No top level `BooleanOp`s

Yeah, great point. I'm kinda sold on the **preemptive type coercion**. I want to think about it a little more before moving forward, but I think that's the direction I'd rather go. We then aren't introducing `Booleans`, rather, we're just saying that `5 < 3 === 0`. I'm ok with that,I think.

**Comment by mrfigg** [comment]
Shall I go ahead and mark #7 ready for review, or should I wait for this to be finalized so it can be added in? I'm fine with either.

**Comment by Morgul** [comment]
Let's go ahead and mark #7 for review, and put this in another merge.

**Comment by Morgul** [comment]
This was closed by #10.
